src/Advection/weno_smoothness.jl

```diff
@@ -0,0 +1,201 @@
+const SS543 = (0.547, -3.882, 4.642, -1.854, 0,  7.043, -17.246, 7.042, 0, 11.003, -9.402, 0, 2.107, 0, 0)
+
+# Coefficients for ninth order WENO
+const SS550 = (1.07918,  -6.49501, 7.58823, -4.11487,  0.86329,  10.20563, -24.62076, 13.58458, -2.88007, 15.21393, -17.04396, 3.64863,  4.82963, -2.08501, 0.22658)
+const SS551 = (0.22658,  -1.40251, 1.65153, -0.88297,  0.18079,   2.42723,  -6.11976,  3.37018, -0.70237,  4.06293,  -4.64976, 0.99213,  1.38563, -0.60871, 0.06908)
+const SS552 = (0.06908,  -0.51001, 0.67923, -0.38947,  0.08209,   1.04963,  -2.99076,  1.79098, -0.38947,  2.31153,  -2.99076, 0.67923,  1.04963, -0.51001, 0.06908)
+const SS553 = (0.06908,  -0.60871, 0.99213, -0.70237,  0.18079,   1.38563,  -4.64976,  3.37018, -0.88297,  4.06293,  -6.11976, 1.65153,  2.42723, -1.40251, 0.22658)
+const SS554 = (0.22658,  -2.08501, 3.64863, -2.88007,  0.86329,   4.82963, -17.04396, 13.58458, -4.11487, 15.21393, -24.62076, 7.58823, 10.20563, -6.49501, 1.07918)
+
+for FT in fully_supported_float_types
+    @eval begin
+        """
+            smoothness_coefficients(::Val{FT}, ::Val{buffer}, ::Val{red_order}, ::Val{stencil})
+
+        Return the coefficients used to calculate the smoothness indicators for the stencil 
+        number `stencil` of a WENO reconstruction of order `buffer * 2 - 1`. The actual order of 
+        reconstruction is restricted by `red_order <= buffer`. The coefficients
+        are ordered in such a way to calculate the smoothness in the following fashion:
+        
+        ```julia
+        buffer  = 4
+        stencil = 0
+        
+        ψ = # The stencil corresponding to S₀ with buffer 4 (7th order WENO)
+        
+        C = smoothness_coefficients(Val(buffer), Val(0))
+        
+        # The smoothness indicator
+        β = ψ[1] * (C[1]  * ψ[1] + C[2] * ψ[2] + C[3] * ψ[3] + C[4] * ψ[4]) + 
+            ψ[2] * (C[5]  * ψ[2] + C[6] * ψ[3] + C[7] * ψ[4]) + 
+            ψ[3] * (C[8]  * ψ[3] + C[9] * ψ[4])
+            ψ[4] * (C[10] * ψ[4])
+        ```
+        
+        In the above case, if `red_order == 2`, then all the coefficients corresponding to ψ[i for i > 2]
+        are zero. This last operation is metaprogrammed in the function `metaprogrammed_smoothness_operation`
+        """
+        # 3rd order WENO, restricted to order 1 (does not matter the restriction order here)
+        @inline smoothness_coefficients(::WENO{2, $FT}, red_order, ::Val{0}) = $(FT.(SS220))
+        @inline smoothness_coefficients(::WENO{2, $FT}, red_order, ::Val{1}) = $(FT.(SS221))
+
+        # 5th order WENO, restricted to orders 3 and 1
+        @inline smoothness_coefficients(::WENO{3, $FT}, red_order, ::Val{0}) =
+                ifelse(red_order == 1, $(FT.(SS310)),    # Order 1                
+                ifelse(red_order == 2, $(FT.(SS320)),    # Order 3                                           
+                                       $(FT.(SS330))))   # Order 5
```

Comment:
do we have to compile these broadcasting operators ?

Reply:
These broadcasting should be computed at precompile time I think, so that the whole $(FT.((coeffs))
is replaced by the actual tuple containing the coefficients with the WENO precision.